### PR TITLE
Add proselint checker

### PR DESCRIPTION
**Pull request by marsam**
Add support for http://proselint.com/

Related: #899

### CODE REVIEW

File: flycheck.el [hunk 1]
```diff
@@ -7841,6 +7843,40 @@ Requires Sphinx 1.2 or newer.  See URL `http://sphinx-doc.org'."
...
 
+(defun flycheck-proselint-parse-errors (output checker buffer)
+  "Parse proselint json output errors from OUTPUT.
+
+CHECKER and BUFFER denoted the CHECKER that returned OUTPUT and
+the BUFFER that was checked respectively.
+
+See URL `http://proselint.com/' for more information about proselint."
+  (let* ((body (json-read-from-string output))
+         (errors (let-alist body .data.errors)))
```
Comment (lunaryorn) [inline flycheck.el]: Can shrink this to a single outer `let-alist`? I.e.

```
(let-alist (json-read-from-string output)
   (mapcar … .data.errors))
```

**Comment by lunaryorn** [comment]
@marsam Great, thank you :+1: Would you mind to add documentation for this syntax checker in `doc/languages.rst`?

**Comment by marsam** [comment]
Updated

**Comment by manuel-uberti** [comment]
Oh yes please! Great idea, thank you. :+1:

**Comment by manuel-uberti** [comment]
Any update on this one?

**Comment by lunaryorn** [comment]
@manuel-uberti I'll merge it when I find the time.

File: flycheck.el [hunk 2]
```diff
@@ -7841,6 +7843,39 @@ Requires Sphinx 1.2 or newer.  See URL `http://sphinx-doc.org'."
...
+                 :checker checker)))
+            .data.errors)))
+
+(flycheck-define-checker proselint
+  "Flycheck checker using Proselint.
+
+See URL `http://proselint.com/'."
+  :command ("proselint" "--json" source)
+  :error-parser flycheck-proselint-parse-errors
+  :modes (text-mode markdown-mode))
```
Comment (mkcode) [inline flycheck.el]: Consider adding `gfm-mode` here as well.

**Comment by mkcode** [comment]
@lunaryorn - I opened a PR on the spacemacs syl20bnr/spacemacs#6012 to introduce something similar. I think that using the `modes` option is correct here, on spacemacs we would want to use the `predicate` option that is hooked up to a `spacemacs toggle`. 

When you get a chance, could you instruct the best way to override and change the modes option to a predicate within the spacemacs layer. Thanks!

**Comment by lunaryorn** [comment]
@mkcode  There's no official way to add or override the `:predicate` of a syntax checker, by design.  A `:predicate` should check for the requirements of a syntax checker; you should not use it to check user options or toggles.   If you'd like to disable a syntax checker then please add it to `flycheck-disabled-checkers`.

**Comment by mkcode** [comment]
> There's no official way to add or override the :predicate of a syntax checker, by design.

Appreciate the feedback. In that case, what arguments could I provide to `define-flycheck-checker so that a checker may be run everywhere (in all modes)? I do not see any examples of this being done.

Appreciate the feedback, perhaps we should continue in https://github.com/syl20bnr/spacemacs/pull/6012

**Comment by lunaryorn** [comment]
@mkcode  I'm afraid that Flycheck doesn't support this case currently, and I'm not sure whether I'd like to add this feature to Flycheck.  I'm not sure what kind of use case your trying to cover, though.

Would you mind to elaborate why you need the checker behave as you said?  Why would you like to *remove* the list of modes?  A syntax checker *without* modes per se is something that I'll remove sooner or later because it makes many things harder than they should be.  IOW, at some point *every* Flycheck checker has to declare their major modes just like every buffer has a major mode.

I'm sorry but we'll have to discuss at this place; I'm not going to join Spacemacs' channels.

**Comment by mkcode** [comment]
> Would you mind to elaborate why you need the checker behave as you said?

Sure. What I want out of the proselint checker is to have it run in all modes. And I also would like the regular modes checkers to run. For example, in Emacs-Lisp mode, I would like (proselint, emacs-lisp, and emacs-lisp-checkdoc) checkers to run. The reason for this, is I would like proselint feedback in my code comments as well.

I seem to have actually made it work actually: https://gist.github.com/d40dfe8fcd0e854b611055f63d276fd3 
In this gist, toggling the `prose-proselint-enabled` to non-nil will make the predicate function return true and enable proselint (in any mode). I then have some magic for adding the mode specific checkers to proselint's `next-checkers` in a `flycheck-before-syntax-check-hook`

> I'm sorry but we'll have to discuss at this place; I'm not going to join Spacemacs' channels.

Understood.

**Comment by lunaryorn** [comment]
Can you use proselint in source code files?!

**Comment by mkcode** [comment]
> Can you use proselint in source code files?!

As far as I can tell, yes - if you provide the filename directly to the `proselint` executable. Providing a dir will filter based on the file extension. Works for me if I explicitly give a file on the command line and more importantly - with flycheck.

screen shot 2016-05-13 at 12 59 30 pm

**Comment by JAremko** [comment]
@lunaryorn 
>Can you use proselint in source code files?!

They have live demo http://proselint.com/write/ you can try it.  I tested .go and .el code and it looks ok except  "curly quotes"  warning is annoying.  Probably it can be disabled with a proselint configuration file.

**Comment by lunaryorn** [comment]
@mkcode I see.  What you'd like to do is not supported by Flycheck yet.  Your hack is clever, but we don't actually give any guarantees that this works; it's a side-effect of how syntax checker selection is implemented currently.

However, I think we just need two changes to Flycheck:

* #836 for a more flexible chaining system which is already in progress in #961.
* More flexible rules for `:modes`, e.g. the ability to not only say `:modes (text-mode …)` but rather something like `:modes (derived text-mode) (derived prog-mode)` or so.

The latter is a rather minor addition to Flycheck as far as I can see, but there's no ticket yet.

@mkcode Do you think you could implement your change to Spacemacs on top of these features?  We could hold back this PR until both requirements are implemented and then update it accordingly.

**Comment by JAremko** [comment]
@lunaryorn   
> something like :modes (derived text-mode) (derived prog-mode) or so.  

That will be really cool :+1:   But it also might be somewhat tricky to implement. For example `web-mode` is derived from `web-mode-prog-mode` - alias for `prog-mode`.  Or `typescript-mode` and `typescript-parent-mode`.  `derived-mode-p` will not detect it as a derived from `prog-mode`  and probably modes that aren't directly derived also won't be detected.

**Comment by lunaryorn** [comment]
@marsam @mkcode @JAremko I'd like to hold this syntax checker back until #961 is merged, and then update this syntax checker right away as outline in my previous comment so that it works seamlessly with all kinds of modes.  Is that ok for you?

**Comment by marsam** [comment]
Sure, no problem :)

**Comment by jpfairbanks** [comment]
I am looking forward to this PR and noticed that issue #961 was superceded by issue #1072. I I just wanted to link to the new issue number here.

**Comment by manuel-uberti** [comment]
Any progress on this? If there is something I can do to help, please tell. :)

**Comment by lunaryorn** [comment]
@manuel-uberti We can't merge this checker before #1072 because it doesn't fit in our current system for chaining.

**Comment by waltervargas** [comment]
Hi @mkcode and @lunaryorn, there is some update on this?, there is something where can I help?

**Comment by cpitclaudel** [comment]
@waltervargas Not much of an update; we're focusing on simple PRs to merge before tackling larger changes like this one (which is somewhat dependent on #1072, which is currently on hold)

**Comment by JAremko** [comment]
Any progress? :roll_eyes:

**Comment by marsam** [comment]
Hi, given that #1072 is taking some time, can this PR be merged? 

P.S. A package with this checker to be included in melpa https://github.com/melpa/melpa/pull/4938, but I think is better to be part of flycheck

**Comment by fmdkdd** [comment]
@marsam Sure, we can merge this.  LGTM.  You just need to rebase on master.

If we include this, then we should probably inform the ["official" plugin](https://github.com/amperser/proselint/blob/master/plugins/flycheck/flycheck-proselint.el) that they don't need to bundle it themselves.

**Comment by mavit** [comment]
> If we include this, then we should probably inform the "official" plugin that they don't need to bundle it themselves.

I'm happy to do that, once this merges.

**Comment by marsam** [comment]
I've made a new PR with these changes rebased in #1304.  Closing this because
having my old fork deleted, GitHub does not allow me to rebase this PR.

File: flycheck.el (hunk 2 above)
Comment (gdkrmr) [reply flycheck.el]: what about `org-mode`?

**Pull request closed without merging**